README.md
# axe

`axe` 是一个基于 `pexpect` 的批量 SSH/SCP 命令行工具，适合一组机器共享相同登录方式时快速执行登录、批量命令和批量文件分发。

它支持：

- 单主机 SSH 登录
- 多主机批量执行命令
- 多主机批量 SCP
- 环境变量和命令行参数覆盖连接配置
- 私钥登录
- `--dry-run` 预览
- `--jobs N` 并发执行

从当前版本开始，`pexpect` 作为外部依赖提供，不再 vendoring 到仓库中。

## Requirements

- `python3`
- 本机可用的 `ssh` 和 `scp`

## Install

直接在仓库里运行：

```bash
./install.sh
```

这会把当前仓库中的 `axe` 写入 `~/.bashrc` 和 `~/.zshrc` 的 alias。

也可以安装成标准命令行包：

```bash
python3 -m pip install .
axe --help
```

如果你在系统包环境中安装 `.deb` 或 `.rpm`，对应包会声明 `python3-pexpect` 依赖，而不是把第三方源码直接打进项目源码目录。

如果你需要在离线环境里部署，也可以使用发布流程生成的离线 `tar.gz` 包。它会把 Python 依赖一起打进去，但目标机器仍需要可用的 `python3`。

## Release

仓库现在包含 GitHub Actions：

- `CI`：在 `push` 和 `pull_request` 时运行测试并验证 Python 构建
- `Release`：在推送 `v*` tag 时自动构建发布产物

发布流程会生成：

- Python `sdist`
- Python `wheel`
- `.deb` 包
- `.rpm` 包
- 离线自包含 `tar.gz` 包

Linux 包发布会输出常见架构标签：

- Debian: `amd64`, `arm64`
- RPM: `x86_64`, `aarch64`

版本变化记录见 [CHANGELOG.md](./CHANGELOG.md)。

触发方式示例：

```bash
git tag v0.1.1
git push origin v0.1.1
```

## Quick Start

```bash
axe 1
axe 10.10.1.1
axe 2 3 4 -c 'uptime'
axe 2 3 4 -s './app.tar.gz' '/srv/app/'
```

默认规则：

- `1` 到 `250` 会被解析为 `192.222.1.x`
- 也支持直接传入 IPv4 地址或域名

## Configuration

默认配置来自环境变量：

```bash
export AXE_USER=root
export AXE_PASSWORD='your-password'
export AXE_PORT=22
export AXE_HOST_PREFIX='192.222.1.'
export AXE_CONNECT_TIMEOUT=15
export AXE_IDENTITY_FILE='~/.ssh/id_rsa'
```

说明：

- `AXE_USER`：登录用户名
- `AXE_PASSWORD`：密码登录时使用的密码，不提供时不会自动发送空密码
- `AXE_PORT`：SSH 端口
- `AXE_HOST_PREFIX`：短主机号前缀，例如 `192.222.1.`
- `AXE_CONNECT_TIMEOUT`：等待连接和密码提示的超时时间，单位秒
- `AXE_IDENTITY_FILE`：私钥路径，可选

命令行参数会覆盖环境变量，仅对当前一次执行生效。

## Usage

### SSH 登录

```bash
axe 1
axe example.com
axe --identity '~/.ssh/id_rsa' 12
```

### 批量执行命令

```bash
axe 2 3 4 -c 'hostname'
axe --jobs 4 2 3 4 -c 'uptime'
axe --dry-run 2 3 4 -c 'systemctl status nginx'
```

### 批量拷贝文件

```bash
axe 2 3 4 -s './test'
axe 2 3 4 -s './test' '/home/admin/'
axe --jobs 3 2 3 4 -s './build.tar.gz' '/srv/app/'
```

当 `-s` 只给一个源路径时，会默认拷贝到远端相同目录。

## Runtime Overrides

下面这些参数可以覆盖默认连接配置：

```bash
axe --user admin --port 2222 --host-prefix '10.20.30.' --timeout 10 12 -c 'hostname'
```

支持的覆盖参数：

- `--user`
- `--password`
- `--port`
- `--host-prefix`
- `--timeout`
- `--identity`
- `--dry-run`
- `--jobs`

## Common Scenarios

### 登录单台短主机号机器

```bash
axe 12
```

等价于连接：

```text
ssh $AXE_USER@192.222.1.12 -p $AXE_PORT
```

### 给一组机器执行同一条巡检命令

```bash
axe 2 3 4 5 -c 'uptime'
```

### 先预览，再真正执行批量命令

```bash
axe --dry-run 2 3 4 -c 'df -h'
axe 2 3 4 -c 'df -h'
```

### 通过私钥连接一组机器

```bash
axe --identity '~/.ssh/id_rsa' 2 3 4 -c 'hostname'
```

### 并发分发构建产物

```bash
axe --jobs 4 2 3 4 5 -s './release.tar.gz' '/srv/app/'
```

## Notes

- `--dry-run` 只输出将要执行的目标和参数，不会真的连接远端。
- `--jobs N` 只作用于批量 `-c` 和 `-s` 场景。
- 批量执行结束后会输出成功/失败汇总。
- 如果远端仍要求密码，工具会继续按当前配置尝试密码交互。
- 不再提供任何内置默认密码，需要显式设置 `AXE_PASSWORD` 或传入 `--password`。

## FAQ

### 为什么 `axe 12` 会连到 `192.222.1.12`？

因为默认会把 `1` 到 `250` 的短主机号拼接到 `AXE_HOST_PREFIX` 后面。默认前缀是 `192.222.1.`。

### 如果我想换成别的网段前缀怎么办？

设置环境变量或命令行参数：

```bash
export AXE_HOST_PREFIX='10.20.30.'
axe 12
```

或者：

```bash
axe --host-prefix '10.20.30.' 12
```

### 如果我不想真的执行，只想看会发生什么？

使用 `--dry-run`：

```bash
axe --dry-run 2 3 4 -c 'systemctl restart nginx'
```

### 如果批量机器很多，执行太慢怎么办？

可以提高并发数：

```bash
axe --jobs 8 2 3 4 5 6 7 8 9 -c 'uptime'
```

### 支持密码和私钥混用吗？

可以。配置了 `--identity` 或 `AXE_IDENTITY_FILE` 时会优先带上私钥；如果远端仍然要求密码，工具会继续处理密码交互。


CHANGELOG.md
# Changelog

## 0.1.1

- Replaced vendored `pexpect` and `ptyprocess` copies with external package dependencies.
- Added an offline self-contained bundle artifact to the release pipeline.
- Tightened release automation and package dependency handling.

## 0.1.0

- Ported the tool to Python 3 and stabilized terminal handling.
- Added regression tests and cleaned Python build artifacts from the repository.
- Improved SSH/SCP behavior for batch command execution and file copy workflows.
- Added configurable host prefix, connection timeout, identity file support, dry-run mode, and batch concurrency.
- Packaged the project as an installable Python CLI with `pyproject.toml`.
- Added GitHub Actions for CI, release builds, and Linux package generation.
- Reworked the README with installation, usage, scenarios, and FAQ sections.

## Before 0.1.0

- Early repository history contains the original shell-oriented utility implementation and incremental fixes.


tests/test_axe.py
import os
import tempfile
import unittest
from pathlib import Path
from unittest import mock

from axe_cli import __version__
from axe_cli import cli as axe


class AxeTests(unittest.TestCase):
    def test_resolve_host_accepts_short_number(self):
        self.assertEqual(axe.resolve_host("12"), "192.222.1.12")

    def test_resolve_host_uses_configured_prefix(self):
        with mock.patch.object(axe, "HOST_PREFIX", "10.20.30."):
            self.assertEqual(axe.resolve_host("12"), "10.20.30.12")

    def test_connect_timeout_comes_from_environment_on_reload(self):
        with mock.patch.dict(os.environ, {"AXE_CONNECT_TIMEOUT": "21"}, clear=False):
            import importlib

            reloaded = importlib.reload(axe)
        self.assertEqual(reloaded.CONNECT_TIMEOUT, 21)

    def test_password_defaults_to_none_when_not_configured(self):
        with mock.patch.dict(os.environ, {}, clear=True):
            import importlib

            reloaded = importlib.reload(axe)
        self.assertIsNone(reloaded.PASSWORD)

    def test_parse_options_supports_cli_overrides(self):
        options = axe.parse_options(
            ["--user", "admin", "--port", "2222", "--host-prefix", "10.0.0.", "--timeout", "9", "--identity", "~/.ssh/id_rsa", "--dry-run", "12"]
        )
        self.assertEqual(options.user, "admin")
        self.assertEqual(options.port, "2222")
        self.assertEqual(options.host_prefix, "10.0.0.")
        self.assertEqual(options.connect_timeout, 9)
        self.assertEqual(options.identity_file, "~/.ssh/id_rsa")
        self.assertTrue(options.dry_run)
        self.assertEqual(options.hosts, ["12"])

    def test_parse_options_requires_values(self):
        with self.assertRaisesRegex(axe.ArgumentParserError, "expected one argument"):
            axe.parse_options(["--port"])

    def test_parse_options_supports_jobs(self):
        options = axe.parse_options(["--jobs", "3", "2", "3", "-c", "hostname"])
        self.assertEqual(options.jobs, 3)
        self.assertEqual(options.hosts, ["2", "3"])
        self.assertEqual(options.command, "hostname")

    def test_apply_runtime_options_updates_globals(self):
        original = (axe.USER, axe.PASSWORD, axe.PORT, axe.HOST_PREFIX, axe.CONNECT_TIMEOUT, axe.IDENTITY_FILE, axe.DRY_RUN, axe.JOBS)
        try:
            axe.apply_runtime_options(
                axe.parse_options(["--user", "admin", "--password", "pw", "--port", "2222", "--host-prefix", "10.0.0.", "--timeout", "8", "--identity", "~/.ssh/id_rsa", "--dry-run", "--jobs", "4"])
            )
            self.assertEqual(axe.USER, "admin")
            self.assertEqual(axe.PASSWORD, "pw")
            self.assertEqual(axe.PORT, "2222")
            self.assertEqual(axe.HOST_PREFIX, "10.0.0.")
            self.assertEqual(axe.CONNECT_TIMEOUT, 8)
            self.assertEqual(axe.IDENTITY_FILE, "~/.ssh/id_rsa")
            self.assertTrue(axe.DRY_RUN)
            self.assertEqual(axe.JOBS, 4)
        finally:
            axe.USER, axe.PASSWORD, axe.PORT, axe.HOST_PREFIX, axe.CONNECT_TIMEOUT, axe.IDENTITY_FILE, axe.DRY_RUN, axe.JOBS = original

    def test_resolve_host_accepts_ipv4(self):
        self.assertEqual(axe.resolve_host("10.0.0.8"), "10.0.0.8")

    def test_resolve_host_rejects_invalid_value(self):
        with self.assertRaises(ValueError):
            axe.resolve_host("999")

    def test_normalize_destination_defaults_to_source_directory(self):
        with tempfile.TemporaryDirectory() as temp_dir:
            file_path = Path(temp_dir) / "demo.txt"
            file_path.write_text("demo", encoding="utf-8")
            self.assertEqual(axe.normalize_destination(str(file_path), None), temp_dir + "/")

    def test_normalize_destination_maps_home_directory_to_tilde(self):
        with mock.patch.dict(os.environ, {"HOME": "/tmp/demo-home"}, clear=False):
            self.assertEqual(axe.normalize_destination("demo.txt", "/tmp/demo-home"), "~")

    def test_main_reports_missing_copy_source(self):
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.main(["1", "-s", "missing-file"])
        self.assertEqual(exit_code, 1)
        mocked_print.assert_called_with("Failed. No such file or directory: missing-file")

    def test_getwinsize_falls_back_without_tty(self):
        with mock.patch.object(axe.sys.stdout, "fileno", side_effect=OSError):
            self.assertEqual(axe.getwinsize(), (24, 80))

    def test_help_flag_prints_help(self):
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.main(["--help"])
        self.assertEqual(exit_code, 0)
        mocked_print.assert_called_once_with(axe.HELP, end="")

    def test_version_flag_prints_version(self):
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.main(["--version"])
        self.assertEqual(exit_code, 0)
        mocked_print.assert_called_once_with(__version__)

    def test_main_applies_cli_overrides_before_running(self):
        original = (axe.USER, axe.PASSWORD, axe.PORT, axe.HOST_PREFIX, axe.CONNECT_TIMEOUT, axe.IDENTITY_FILE, axe.DRY_RUN, axe.JOBS)
        try:
            with mock.patch.object(axe, "astute_ssh") as ssh_mock:
                exit_code = axe.main(
                    ["--user", "admin", "--port", "2222", "--host-prefix", "10.0.0.", "--timeout", "8", "--identity", "~/.ssh/id_rsa", "--dry-run", "12"]
                )
            self.assertEqual(exit_code, 0)
            ssh_mock.assert_called_once_with("12")
            self.assertEqual(axe.USER, "admin")
            self.assertEqual(axe.PORT, "2222")
            self.assertEqual(axe.HOST_PREFIX, "10.0.0.")
            self.assertEqual(axe.CONNECT_TIMEOUT, 8)
            self.assertEqual(axe.IDENTITY_FILE, "~/.ssh/id_rsa")
            self.assertTrue(axe.DRY_RUN)
            self.assertEqual(axe.JOBS, 1)
        finally:
            axe.USER, axe.PASSWORD, axe.PORT, axe.HOST_PREFIX, axe.CONNECT_TIMEOUT, axe.IDENTITY_FILE, axe.DRY_RUN, axe.JOBS = original

    def test_main_reports_missing_option_value(self):
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.main(["--timeout"])
        self.assertEqual(exit_code, 1)
        mocked_print.assert_called_once_with("Failed: argument --timeout: expected one argument")

    def test_main_rejects_invalid_jobs(self):
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.main(["--jobs", "0", "2", "-c", "hostname"])
        self.assertEqual(exit_code, 1)
        mocked_print.assert_called_once_with("Failed: argument --jobs: must be at least 1")

    def test_main_rejects_command_without_hosts(self):
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.main(["-c", "hostname"])
        self.assertEqual(exit_code, 1)
        mocked_print.assert_called_with("Failed. No host specified for command execution.")

    def test_main_rejects_scp_without_hosts(self):
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.main(["-s", "demo.txt"])
        self.assertEqual(exit_code, 1)
        mocked_print.assert_called_with("Failed. No host specified for SCP.")

    def test_print_batch_summary_without_failures(self):
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.print_batch_summary("command execution", ["2", "3"], [])
        self.assertEqual(exit_code, 0)
        mocked_print.assert_any_call("Completed command execution on all hosts.")
        mocked_print.assert_any_call("Summary: total=2, succeeded=2, failed=0")

    def test_print_batch_summary_with_failures(self):
        failures = [{"host": "2", "error": "timeout"}, {"host": "4", "error": "auth failed"}]
        with mock.patch("builtins.print") as mocked_print:
            exit_code = axe.print_batch_summary("SCP", ["2", "3", "4"], failures)
        self.assertEqual(exit_code, 1)
        mocked_print.assert_any_call("Completed SCP with failures on: 2, 4")
        mocked_print.assert_any_call("Summary: total=3, succeeded=1, failed=2")
        mocked_print.assert_any_call("  2: timeout")
        mocked_print.assert_any_call("  4: auth failed")

    def test_run_command_continues_after_host_failure(self):
        with mock.patch.object(axe, "astute_ssh", side_effect=[RuntimeError("timeout"), None]) as ssh_mock, \
             mock.patch("builtins.print") as mocked_print:
            exit_code = axe.run_command(["2", "3"], ["hostname"])
        self.assertEqual(exit_code, 1)
        self.assertEqual(ssh_mock.call_count, 2)
        mocked_print.assert_any_call("Completed command execution with failures on: 2")

    def test_run_scp_continues_after_host_failure(self):
        with tempfile.TemporaryDirectory() as temp_dir:
            file_path = Path(temp_dir) / "demo.txt"
            file_path.write_text("demo", encoding="utf-8")
            with mock.patch.object(axe, "astute_scp", side_effect=[RuntimeError("timeout"), None]) as scp_mock, \
                 mock.patch("builtins.print") as mocked_print:
                exit_code = axe.run_scp(["2", "3"], [str(file_path)])
        self.assertEqual(exit_code, 1)
        self.assertEqual(scp_mock.call_count, 2)
        mocked_print.assert_any_call("Completed SCP with failures on: 2")

    def test_run_batch_uses_jobs_setting(self):
        original_jobs = axe.JOBS
        try:
            axe.JOBS = 2
            seen = []

            def worker(host):
                seen.append(host)

            failures = axe.run_batch(["2", "3"], worker)
            self.assertEqual(failures, [])
            self.assertCountEqual(seen, ["2", "3"])
        finally:
            axe.JOBS = original_jobs

    def test_ssh_command_mode_waits_for_completion(self):
        child = mock.Mock()
        with mock.patch.object(axe.pexpect, "spawn", return_value=child) as spawn_mock, \
             mock.patch.object(axe, "expect_and_send_password") as expect_mock, \
             mock.patch.object(axe, "wait_for_child") as wait_mock, \
             mock.patch.object(axe, "interact_with_child") as interact_mock:
            axe.ssh("root", "secret", "10.0.0.8", command="hostname")

        spawn_mock.assert_called_once()
        expect_mock.assert_called_once_with(child, "secret")
        wait_mock.assert_called_once_with(child)
        interact_mock.assert_not_called()
        self.assertIs(child.logfile_read, axe.sys.stdout)

    def test_scp_waits_for_completion(self):
        child = mock.Mock()
        with mock.patch.object(axe.pexpect, "spawn", return_value=child) as spawn_mock, \
             mock.patch.object(axe, "expect_and_send_password") as expect_mock, \
             mock.patch.object(axe, "wait_for_child") as wait_mock:
            axe.scp("root", "secret", "10.0.0.8", "demo.txt", destination="/tmp")

        spawn_mock.assert_called_once()
        expect_mock.assert_called_once_with(child, "secret")
        wait_mock.assert_called_once_with(child)
        self.assertIs(child.logfile_read, axe.sys.stdout)

    def test_ssh_includes_identity_file_when_configured(self):
        child = mock.Mock()
        with mock.patch.object(axe.pexpect, "spawn", return_value=child) as spawn_mock, \
             mock.patch.object(axe, "expect_and_send_password"), \
             mock.patch.object(axe, "wait_for_child"):
            axe.ssh("root", "secret", "10.0.0.8", command="hostname", identity_file="~/.ssh/id_rsa")
        self.assertIn("-i", spawn_mock.call_args.kwargs["args"])
        self.assertIn(os.path.expanduser("~/.ssh/id_rsa"), spawn_mock.call_args.kwargs["args"])

    def test_expect_and_send_password_uses_configured_timeout(self):
        child = mock.Mock()
        child.expect.return_value = 0
        with mock.patch.object(axe, "CONNECT_TIMEOUT", 27):
            axe.expect_and_send_password(child, "secret")
        child.expect.assert_called_once_with(
            [r"[Pp]assword:", r"Permission denied", r"Connection refused", r"No route to host", axe.pexpect.EOF, axe.pexpect.TIMEOUT],
            timeout=27,
        )
        child.sendline.assert_called_once_with("secret")

    def test_expect_and_send_password_requires_configured_password(self):
        child = mock.Mock()
        child.expect.return_value = 0
        with self.assertRaisesRegex(RuntimeError, "no password is configured"):
            axe.expect_and_send_password(child, None)
        child.close.assert_called_once()

    def test_expect_and_send_password_classifies_auth_failure(self):
        child = mock.Mock()
        child.expect.return_value = 1
        with self.assertRaisesRegex(RuntimeError, "Authentication failed"):
            axe.expect_and_send_password(child, "secret")

    def test_expect_and_send_password_classifies_connection_refused(self):
        child = mock.Mock()
        child.expect.return_value = 2
        with self.assertRaisesRegex(RuntimeError, "Connection refused"):
            axe.expect_and_send_password(child, "secret")

    def test_wait_for_child_classifies_remote_close(self):
        child = mock.Mock()
        child.expect.return_value = 1
        with self.assertRaisesRegex(RuntimeError, "Connection closed by remote host"):
            axe.wait_for_child(child)
        child.close.assert_called_once()

    def test_wait_for_child_classifies_timeout(self):
        child = mock.Mock()
        child.expect.return_value = 3
        with self.assertRaisesRegex(RuntimeError, "Timed out waiting for remote command to finish"):
            axe.wait_for_child(child)
        child.close.assert_called_once()

    def test_astute_ssh_dry_run_prints_command(self):
        original = axe.DRY_RUN
        try:
            axe.DRY_RUN = True
            with mock.patch("builtins.print") as mocked_print:
                axe.astute_ssh("12", "hostname")
            mocked_print.assert_called_once_with("DRY RUN ssh {}@{} port={} command={!r}".format(axe.USER, "192.222.1.12", axe.PORT, "hostname"))
        finally:
            axe.DRY_RUN = original

    def test_astute_scp_dry_run_prints_transfer(self):
        original = axe.DRY_RUN
        try:
            axe.DRY_RUN = True
            with tempfile.TemporaryDirectory() as temp_dir:
                file_path = Path(temp_dir) / "demo.txt"
                file_path.write_text("demo", encoding="utf-8")
                with mock.patch("builtins.print") as mocked_print:
                    axe.astute_scp("12", str(file_path))
            mocked_print.assert_called_once()
            self.assertIn("DRY RUN scp", mocked_print.call_args.args[0])
        finally:
            axe.DRY_RUN = original


if __name__ == "__main__":
    unittest.main()


scripts/write_nfpm_config.py
#!/usr/bin/env python3

import sys
from pathlib import Path


def main() -> int:
    if len(sys.argv) != 5:
        print("Usage: write_nfpm_config.py <output> <version> <arch> <package-format>", file=sys.stderr)
        return 1

    output, version, arch, package_format = sys.argv[1:]
    config = f"""name: axe-cli
arch: {arch}
platform: linux
version: {version}
section: utils
priority: optional
maintainer: axe contributors
description: Batch SSH and SCP helper built on pexpect
homepage: https://github.com/axeprpr/axe
contents:
  - src: ./dist/package-root/usr/lib/axe/
    dst: /usr/lib/axe
  - src: ./dist/package-root/usr/bin/axe
    dst: /usr/bin/axe
    file_info:
      mode: 0755
depends:
  - python3
  - python3-pexpect
"""
    if package_format == "rpm":
        config += "rpm:\n  compression: gzip\n"

    Path(output).write_text(config, encoding="utf-8")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())


axe_cli/cli.py
import argparse
import fcntl
import os
import platform
import re
import signal
import struct
import sys
import termios
from concurrent.futures import ThreadPoolExecutor

import pexpect

from axe_cli import __version__


HELP = """help:
axe 1 ==> ssh root@192.222.1.1 (use root password of astute)
axe 10.10.1.1 ==> ssh root@10.10.1.1 (use root password of astute)
axe 2 3 4 -c 'ls -lrt' ==> run command on host002/3/4 and show result
axe 2 3 4 -s './test' '/home/astute' ==> scp file to host002/3/4 on purpose
axe 2 3 4 -s './test' ==> scp file to host002/3/4 to the same place
"""

USER = os.environ.get("AXE_USER", "root")
PASSWORD = os.environ.get("AXE_PASSWORD")
PORT = os.environ.get("AXE_PORT", "22")
HOST_PREFIX = os.environ.get("AXE_HOST_PREFIX", "192.222.1.")
CONNECT_TIMEOUT = int(os.environ.get("AXE_CONNECT_TIMEOUT", "15"))
IDENTITY_FILE = os.environ.get("AXE_IDENTITY_FILE")
DRY_RUN = False
JOBS = 1
CURRENT_CHILD = None


def sigwinch_passthrough(sig, data):
    del sig, data
    if CURRENT_CHILD is None:
        return
    rows, cols = getwinsize()
    CURRENT_CHILD.setwinsize(rows, cols)


def getwinsize():
    """Return the current terminal size as (rows, cols)."""
    if "TIOCGWINSZ" in dir(termios):
        ioctl_code = termios.TIOCGWINSZ
    else:
        ioctl_code = 1074295912
    packed = struct.pack("HHHH", 0, 0, 0, 0)
    try:
        data = fcntl.ioctl(sys.stdout.fileno(), ioctl_code, packed)
    except (OSError, AttributeError):
        return (24, 80)
    return struct.unpack("HHHH", data)[0:2]


def is_domain(domain):
    domain_regex = re.compile(
        r"(?:[A-Z0-9_](?:[A-Z0-9-_]{0,247}[A-Z0-9])?\.)+(?:[A-Z]{2,6}|[A-Z0-9-]{2,}(?<!-))\Z",
        re.IGNORECASE,
    )
    return bool(domain_regex.match(domain))


def is_ipv4(address):
    ipv4_regex = re.compile(
        r"^(?:25[0-5]|2[0-4]\d|[0-1]?\d?\d)(?:\.(?:25[0-5]|2[0-4]\d|[0-1]?\d?\d)){3}$",
        re.IGNORECASE,
    )
    return bool(ipv4_regex.match(address))


def resolve_host(value):
    host = str(value).strip()
    if not host:
        raise ValueError("Host is empty.")
    if host.isdigit():
        number = int(host)
        if 1 <= number <= 250:
            return "{}{}".format(HOST_PREFIX, number)
        raise ValueError("Host number {} is out of range.".format(host))
    if is_domain(host) or is_ipv4(host):
        return host
    raise ValueError("Unsupported host '{}'.".format(host))


def normalize_destination(file_path, destination):
    if destination:
        if os.path.abspath(os.path.expanduser(destination)).lower() == os.path.expanduser("~").lower():
            return "~"
        return destination

    target_dir = os.path.dirname(os.path.abspath(file_path))
    if not target_dir.endswith("/"):
        target_dir += "/"
    return target_dir


class ArgumentParserError(ValueError):
    pass


class AxeArgumentParser(argparse.ArgumentParser):
    def error(self, message):
        raise ArgumentParserError(message)


def build_parser():
    parser = AxeArgumentParser(add_help=False, prog="axe", description="Batch SSH and SCP helper")
    parser.add_argument("hosts", nargs="*", help="Short host numbers, IPv4 addresses, or domain names")
    parser.add_argument("-h", "--help", action="store_true", dest="show_help")
    parser.add_argument("--version", action="store_true")
    parser.add_argument("-c", "--command", dest="command")
    parser.add_argument("-s", "--scp", nargs="+", dest="scp_args")
    parser.add_argument("--user", default=USER)
    parser.add_argument("--password", default=PASSWORD)
    parser.add_argument("--port", default=PORT)
    parser.add_argument("--host-prefix", dest="host_prefix", default=HOST_PREFIX)
    parser.add_argument("--timeout", dest="connect_timeout", type=int, default=CONNECT_TIMEOUT)
    parser.add_argument("--identity", dest="identity_file", default=IDENTITY_FILE)
    parser.add_argument("--dry-run", action="store_true", default=DRY_RUN)
    parser.add_argument("--jobs", type=int, default=1)
    return parser


def parse_options(argv):
    parser = build_parser()
    namespace = parser.parse_intermixed_args(argv)
    if namespace.jobs < 1:
        raise ArgumentParserError("argument --jobs: must be at least 1")
    if namespace.command and namespace.scp_args:
        raise ArgumentParserError("argument -s/--scp: not allowed with argument -c/--command")
    return namespace


def apply_runtime_options(options):
    global USER, PASSWORD, PORT, HOST_PREFIX, CONNECT_TIMEOUT, IDENTITY_FILE, DRY_RUN, JOBS

    USER = options.user
    PASSWORD = options.password
    PORT = str(options.port)
    HOST_PREFIX = options.host_prefix
    CONNECT_TIMEOUT = int(options.connect_timeout)
    IDENTITY_FILE = options.identity_file
    DRY_RUN = bool(options.dry_run)
    JOBS = int(options.jobs)


def expect_and_send_password(child, password):
    match_index = child.expect(
        [r"[Pp]assword:", r"Permission denied", r"Connection refused", r"No route to host", pexpect.EOF, pexpect.TIMEOUT],
        timeout=CONNECT_TIMEOUT,
    )
    if match_index == 0:
        if not password:
            child.close()
            raise RuntimeError("Password prompt received but no password is configured.")
        child.sendline(password)
        return
    if match_index == 1:
        raise RuntimeError("Authentication failed.")
    if match_index == 2:
        raise RuntimeError("Connection refused.")
    if match_index == 3:
        raise RuntimeError("No route to host.")
    if match_index == 4:
        return
    raise RuntimeError("Timed out waiting for password prompt.")


def wait_for_child(child):
    match_index = child.expect([r"Permission denied", r"Connection closed by remote host", pexpect.EOF, pexpect.TIMEOUT])
    if match_index == 0:
        child.close()
        raise RuntimeError("Authentication failed.")
    if match_index == 1:
        child.close()
        raise RuntimeError("Connection closed by remote host.")
    if match_index == 3:
        child.close()
        raise RuntimeError("Timed out waiting for remote command to finish.")

    child.close()
    if child.exitstatus not in (0, None):
        raise RuntimeError("Command exited with status {}.".format(child.exitstatus))
    if child.signalstatus is not None:
        raise RuntimeError("Command terminated by signal {}.".format(child.signalstatus))


def print_batch_summary(action, hosts, failures):
    total = len(hosts)
    failed = len(failures)
    succeeded = total - failed

    if not failures:
        print("Completed {} on all hosts.".format(action))
        print("Summary: total={}, succeeded={}, failed=0".format(total, succeeded))
        return 0

    failed_hosts = ", ".join(item["host"] for item in failures)
    print("Completed {} with failures on: {}".format(action, failed_hosts))
    print("Summary: total={}, succeeded={}, failed={}".format(total, succeeded, failed))
    for item in failures:
        print("  {}: {}".format(item["host"], item["error"]))
    return 1


def interact_with_child(child):
    global CURRENT_CHILD

    old_handler = None
    CURRENT_CHILD = child
    if "Microsoft" not in platform.platform():
        old_handler = signal.getsignal(signal.SIGWINCH)
        signal.signal(signal.SIGWINCH, sigwinch_passthrough)
        rows, cols = getwinsize()
        child.setwinsize(rows, cols)

    try:
        child.interact()
    finally:
        CURRENT_CHILD = None
        if old_handler is not None:
            signal.signal(signal.SIGWINCH, old_handler)


def ssh(user, password, ip, port=PORT, command="", identity_file=None):
    args = [
        "-o",
        "UserKnownHostsFile=/dev/null",
        "-o",
        "StrictHostKeyChecking=no",
        "-p",
        str(port),
    ]
    if identity_file:
        args.extend(["-i", os.path.expanduser(identity_file)])
    args.append("{}@{}".format(user, ip))
    if command:
        args.append(command)

    child = pexpect.spawn(
        "ssh",
        args=args,
        env=dict(os.environ, TERM=os.environ.get("TERM", "xterm-256color")),
        encoding="utf-8",
    )
    expect_and_send_password(child, password)
    if command:
        child.logfile_read = sys.stdout
        wait_for_child(child)
    else:
        interact_with_child(child)


def scp(user, password, ip, file_path, port=PORT, destination=None, identity_file=None):
    target = normalize_destination(file_path, destination)
    args = [
        "-o",
        "UserKnownHostsFile=/dev/null",
        "-o",
        "StrictHostKeyChecking=no",
        "-P",
        str(port),
    ]
    if identity_file:
        args.extend(["-i", os.path.expanduser(identity_file)])
    args.extend([
        "-r",
        file_path,
        "{}@{}:{}".format(user, ip, target),
    ])
    child = pexpect.spawn("scp", args=args, env=dict(os.environ), encoding="utf-8")
    expect_and_send_password(child, password)
    child.logfile_read = sys.stdout
    wait_for_child(child)


def astute_ssh(ip, command=""):
    if DRY_RUN:
        target = resolve_host(ip)
        print("DRY RUN ssh {}@{} port={} command={!r}".format(USER, target, PORT, command))
        return
    ssh(USER, PASSWORD, resolve_host(ip), command=command, identity_file=IDENTITY_FILE)


def astute_scp(ip, file_path, destination=None):
    if DRY_RUN:
        target = resolve_host(ip)
        print("DRY RUN scp {} -> {}@{}:{} port={}".format(file_path, USER, target, normalize_destination(file_path, destination), PORT))
        return
    scp(USER, PASSWORD, resolve_host(ip), file_path, destination=destination, identity_file=IDENTITY_FILE)


def run_batch(hosts, worker):
    failures = []
    if JOBS <= 1 or len(hosts) <= 1:
        for host in hosts:
            try:
                worker(host)
            except Exception as exc:
                print("Failed: {}".format(exc))
                failures.append({"host": str(host), "error": str(exc)})
        return failures

    with ThreadPoolExecutor(max_workers=JOBS) as executor:
        futures = {host: executor.submit(worker, host) for host in hosts}
        for host, future in futures.items():
            try:
                future.result()
            except Exception as exc:
                print("Failed: {}".format(exc))
                failures.append({"host": str(host), "error": str(exc)})
    return failures


def run_scp(hosts, command):
    if not hosts:
        print("Failed. No host specified for SCP.")
        return 1
    if len(command) == 0 or len(command) > 2:
        print(HELP, end="")
        return 1

    file_path = command[0]
    if not os.path.exists(file_path):
        print("Failed. No such file or directory: {}".format(file_path))
        return 1

    for host in hosts:
        if len(command) == 1:
            destination = None
            print("\033[41;1m SCP {} to host-{} to the same place.\033[0m".format(file_path, host))
        else:
            destination = normalize_destination(file_path, command[1])
            print("\033[41;1m SCP {} to host-{} to {}\033[0m".format(file_path, host, destination))
        print("----------------------------")
    failures = run_batch(hosts, lambda host: astute_scp(host, file_path, destination=destination))
    return print_batch_summary("SCP", hosts, failures)


def run_command(hosts, command):
    if not hosts:
        print("Failed. No host specified for command execution.")
        return 1
    if len(command) != 1:
        print(HELP, end="")
        return 1

    for host in hosts:
        print("\033[41;1m Run command {} in host-{}\033[0m".format(command[0], host))
        print("----------------------------")
    failures = run_batch(hosts, lambda host: astute_ssh(host, command[0]))
    return print_batch_summary("command execution", hosts, failures)


def main(argv=None):
    parameter = list(sys.argv[1:] if argv is None else argv)
    try:
        options = parse_options(parameter)
        apply_runtime_options(options)
    except ArgumentParserError as exc:
        print("Failed: {}".format(exc))
        return 1

    if options.version:
        print(__version__)
        return 0

    if options.show_help or (not options.hosts and not options.command and not options.scp_args):
        print(HELP, end="")
        return 0

    if len(options.hosts) == 1 and not options.command and not options.scp_args:
        try:
            astute_ssh(options.hosts[0])
        except Exception as exc:
            print("Failed: {}".format(exc))
            return 1
        return 0

    if options.scp_args is not None and options.command is None:
        return run_scp(options.hosts, options.scp_args)

    if options.command is not None and options.scp_args is None:
        return run_command(options.hosts, [options.command])

    print(HELP, end="")
    return 1
